### DISCUSSION TITLE
Adding extra directories outside root_dir ?

### DISCUSSION THREAD

**Discussion by gitced** [cesanta/mongoose#2591] [Ask questions, get answers] [answered]
Hello,
I need to create several mongoose servers that all share a common directory (containing lots of png files, hence the need to share).
Is that possible ? Can i declare one or more directories to be served outside root_dir ?
Thanks
Cedric

**Comment by scaprile** [comment, marked as answer]
Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.

Why do you _need_ "several mongoose servers" in the first place ? What are you talking about ? Mongoose struct mgr instances ? Actual hosts running mongoose with some file system mounted in all of them ?
Mongoose does not know about serving files, you actually write your server event handler. Will you plan to use which function ?
https://mongoose.ws/documentation/#mg_http_serve_file ?
https://mongoose.ws/documentation/#mg_http_serve_dir ?
What do you mean by "outside root_dir" ? What is "root_dir" for you, the element in `struct mg_http_serve_opts` ? https://mongoose.ws/documentation/#struct-mg_http_serve_opts
And what would "outside" mean ? Any file from the root going down the tree will be served by mg_http_serve_dir. Of course, due to security issues, Mongoose will reject double dots in URL paths, _down the tree_ only.
Any file with a specific path will be served by mg_http_serve_file as long as you pass it the desired path
Mongoose does not know your path is an NFS/SMB/whatever mount, it will just use the underlying file system to traverse directories and read files, nothing prevents you from mounting your whatever in a subdirectory of your root directory; nothing prevents you from having as many mg_http_serve_dir() calls to serve as many URIs as you want; and nothing prevents you from indicating several directories in your opts.root_dir, as the docs clearly say:
>    opts - Serve options. Note that opts.root_dir can optionally accept extra comma-separated uri=path pairs, see example below
>    `opts.root_dir = "/var/www,/conf=/etc";  // Serve /var/www. URIs starting with /conf are served from /etc`

You just put your files where you want them to be and instruct Mongoose to serve from that directory and down the tree.
The only caveat is that your files have to be "stat-able", you can't serve /proc, for example

**Reply by gitced** [reply]
Thanks a lot for your answer.
I just had to filter the address with mg_http_match_uri and define the root_dir (document root) accordingly.
Sorry about the stupid question and thanks again for your lighting fast response.
Cedric

**Reply by scaprile** [reply]
No question is stupid, sometimes the other end can't see what's in your head and asks a loooooooooot of questions...
